### ISSUE TITLE
"prefer-reflect" rule activated by using d3's `call`

### ISSUE THREAD

**Issue by jameswomack** [eslint/eslint#6387]
**What version of ESLint are you using?**
v2.12.0

**What parser (default, Babel-ESLint, etc.) are you using?**
babel-eslint

**Please show your full configuration:**
```json
{
  "parser": "babel-eslint",
  "plugins": [ "react" ],
  "env": {
    "es6": true,
    "worker": true,
    "node": true,
    "browser": true,
    "builtin": true
  },
  "parserOptions": {
    "ecmaVersion": 7,
    "sourceType": "module",
    "ecmaFeatures": {
      "jsx": true,
      "objectLiteralDuplicateProperties": true,
      "destructuring": true,
      "defaultParams": true,
      "classes": true,
      "arrowFunctions": true,
      "modules": true
    }
  },
  "globals": { },
  "rules": {
    "block-scoped-var": 2,
    "camelcase": 2,
    "comma-style": [
      2,
      "last"
    ],
    "curly": 0,
    "dot-notation": [
      2,
      {
        "allowKeywords": true
      }
    ],
    "eqeqeq": [
      2,
      "allow-null"
    ],
    "guard-for-in": 2,
    "key-spacing": 0,
    "new-cap": 0,
    "no-bitwise": 0,
    "no-caller": 2,
    "no-cond-assign": [
      2,
      "except-parens"
    ],
    "no-debugger": 2,
    "no-empty": 2,
    "no-eval": 2,
    "no-extend-native": 2,
    "no-extra-parens": 2,
    "no-irregular-whitespace": 2,
    "no-iterator": 2,
    "no-loop-func": 2,
    "no-multi-spaces": 0,
    "no-multi-str": 2,
    "no-new": 2,
    "no-plusplus": 0,
    "no-proto": 2,
    "no-script-url": 2,
    "no-sequences": 2,
    "no-shadow": 2,
    "no-undef": 2,
    "no-unused-expressions": 0,
    "no-unused-vars": 2,
    "no-underscore-dangle": 0,
    "no-with": 2,
    "no-multiple-empty-lines": [
      2, {"max": 3, "maxEOF": 1}
    ],
    "quotes": 0,
    "semi": 0,
    "semi-spacing": 0,
    "strict": 0,
    "valid-typeof": 2,
    "wrap-iife": [
      2,
      "inside"
    ],
    "no-var": 1,
    "prefer-template": 1,
    "prefer-spread": 1,
    "prefer-arrow-callback": 1,
    "prefer-const": 1,
    "prefer-rest-params": 1,
    "prefer-reflect": 1,
    "no-func-assign": 1,
    "no-dupe-class-members": 1,
    "no-class-assign": 1
  }
}
```

**What did you do? Please include the actual source code causing the issue.**
Edited this file: https://gist.github.com/jameswomack/ce5d267758a52fa49395aadef5f9074c

**What did you expect to happen?**
Although I understand why this is happening, I hope  we can arrive at an elegant solution for D3 as it's quite popular.

**What actually happened? Please include the actual, raw output from ESLint.**
```sh
src/analyzeColors.js
  303:9  warning  Avoid using Function.prototype.call, instead use Reflect.apply  prefer-reflect
  308:9  warning  Avoid using Function.prototype.call, instead use Reflect.apply  prefer-reflect
  313:9  warning  Avoid using Function.prototype.call, instead use Reflect.apply  prefer-reflect

✖ 3 problems (0 errors, 3 warnings)
```

**Comment by ilyavolodin** [comment]
You can configure the rule with `/*eslint prefer-reflect: ["error", { exceptions: ["call"] }]*/` ESLint doesn't have library specific checks, and there's no way for us to generally identify that example that you provided should not be flagged.

For the people check it from mobile, example in question is:
```js
/*eslint prefer-reflect: ["error"]*/
svg.append("g")
            .attr("class", "x axis")
            .attr("transform", "translate(0," + (height - margin.top) + ")")
            .call(xAxis);
```

**Comment by jameswomack** [comment]
@ilyavolodin Thanks for letting me know about exceptions w/ prefer-reflect. 

In an ESLint plugin I might target d3 methods by 
1. testing if d3 (or whatever) was imported or required 
2. excluding `call` when it's chained to d3 methods like `attr` (or whatever)
The config implementation could be something like:
`/* eslint prefer-reflect: ["error", { excluded: [{ import : "d3", methods : [ "attr", "append" ] }] }] */`

**Comment by ilyavolodin** [comment]
@jameswomack ESLint core itself is library agnostic. We don't have any rules that are specific to libraries (except JSDoc, which is more of a style than a library). That's what we created plugin framework for. There are a lot of plugins out there that are library-specific and provide rules that are targeting some library. If `exceptions` property doesn't work for your needs, my suggestion would be to create a D3-specific ESLint plugin, clone and modify this rule to detect D3 and not flag those statements. I'm sure there's a bunch more rules that could be created specifically for D3 code as well.

**Comment by jameswomack** [comment]
@ilyavolodin I know what you mean—that's why I typed "In an ESLint **plugin** I might..." 😄 . I'll edit my post to make that more clear. The agnosticism of core is great, as is the plugin framework. It's why I switched from JSHint a couple years ago. 

P.S. I think you do a great job in managing issues on ESLint. It's rare for a repo to have folks that are so willing to respond quickly and in detail.
